### ISSUE TITLE
using -a to set property is not working

### ISSUE THREAD

**Issue by KiaKK** [docker-solr/docker-solr#221]
Hi There,

I am trying to run solr in different environments. data-config.xml file stays the same for all environments, and I use ${datasource.user} as the placeholder and try to pass in the correct username whiling running the container, but that seems not working for me.

The command used to start the container is:
```
docker run --rm --name solrcontainer -p 8983:8983 -v "$PWD/indexes:/opt/solr/server/solr/data" solrimage -t "/opt/solr/server/solr/data" -a "-Ddatasource.user=REALUSERNAME"
```
where solrimage is the image built from dockerfile below:
```
FROM solr:7.7.1

EXPOSE 8983

ADD jars server/solr-webapp/webapp/WEB-INF/lib
COPY ConfigNoData/solr server/solr/

USER root
RUN chown -R solr:solr server/solr
USER solr
```
But after running the command, the ${datasource.user} in data-config.xml was not replaced by REALUSERNAME. Is there anything I did wrong here?

Also I am pretty new to docker. If there are any changes needed to make the dockerfile and docker run command better, please feel free to comment. Thank you!

**Comment by makuk66** [comment]
Couple of comments:
- what does your `data-config.xml` look like? in particular the part where you consume the placeholder
- have you verified that it turns up in the correct place in the container?
- Have you verified the `-Ddatasource.user` shows up in the JVM argument list in the docker container?
- are you following any specific recipe/example/instructions?
- any reason you're on Solr 7 rather than 8? For 8, just mount a volume on `/var/solr/data`
- can you share a full repo with all the bits?
- If not, perhaps try to create a repeatable example for just this issue. It'd be nice to have a complete working DIH example for Solr 8 complete with DB using Docker compose, but I don't have one and it would take a bunch of time to create.

**Comment by makuk66** [comment]
Your question is how to pass in different DB config for different environments. You didn't say what environments or why, but I assume one use-case is to run against a staging environment where the database URL and credentials are different. What I would personally try to do is have separate directories for each environment, each with their own `data-config.xml`. If you want consistency, generate these config files from a single bash script.

What you tried is to use _generic_ JVM properties in the `data-config.xml`, expecting them to be resolved. I don't think that variable resolver processing that template will resolve those (though someone might correct me; it seems obvious and useful). That resolver deals with a special namespaces, see @jamiejackson's [method](https://github.com/docker-solr/docker-solr/issues/68#issuecomment-251782055). 

There are some other alternatives. The FAQ has [this mechanism](https://wiki.apache.org/solr/DataImportHandlerFaq#Is_it_possible_to_use_core_properties_inside_data-config_xml.3F), and the Ref Guide shows how you can use [request parameters](https://lucene.apache.org/solr/guide/8_1/uploading-structured-data-store-data-with-the-data-import-handler.html#dih-request-parameters) (which I think is somewhat dubious).

So let's try the special namespace approach.


Let's fire up a Postgres DB with some data in it. Here's one I found on Docker hub:

```
$ docker run -d --name pg-ds-world aa8y/postgres-dataset:world

$ docker exec -it pg-ds-world bash

$ psql world
              List of relations
 Schema |      Name       | Type  |  Owner   
--------+-----------------+-------+----------
 public | city            | table | postgres
 public | country         | table | postgres
 public | countrylanguage | table | postgres
(3 rows)

world=# \d city
                     Table "public.city"
   Column    |     Type     | Collation | Nullable | Default 
-------------+--------------+-----------+----------+---------
 id          | integer      |           | not null | 
 name        | text         |           | not null | 
 countrycode | character(3) |           | not null | 
 district    | text         |           | not null | 
 population  | integer      |           | not null | 
Indexes:
    "city_pkey" PRIMARY KEY, btree (id)
Referenced by:
    TABLE "country" CONSTRAINT "country_capital_fkey" FOREIGN KEY (capital) REFERENCES city(id)

world=# select name from city limit 2;                                                                                                                                                                         name   
----------
 Kabul
 Qandahar
(2 rows)
```

let's see what its IP address is:

```
$ docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' pg-ds-world  
172.17.0.2
```

and double check we can get there from another container:

```
docker run -it --rm --name pg-ds-world-client aa8y/postgres-dataset:world bash
bash-4.4$ psql -h 172.17.0.2 -U postgres world
Password for user postgres: 
psql (11.1)
Type "help" for help.

world=# \d
              List of relations
 Schema |      Name       | Type  |  Owner   
--------+-----------------+-------+----------
 public | city            | table | postgres
 public | country         | table | postgres
 public | countrylanguage | table | postgres
(3 rows)
```

OK, that's the DB running.


To talk to it from Solr we'll need a JDBC driver.
```
curl -O https://jdbc.postgresql.org/download/postgresql-42.2.6.jar
```
We'll use a Solr DIH `data-config.xml` with explicit values first:

```
<dataConfig>
  <dataSource type="JdbcDataSource"
              driver="com.mysql.jdbc.Driver"
              url="jdbc:mysql://172.17.0.2/world"  
              user="postgresql"     
              password="postgresql"/>
  <document>
    <entity name="id"
            query="select id,name from city">
     <field column="id" name="id"/>
     <field column="name" name="name_str"/>       
    </entity>
  </document>
</dataConfig>
```

Grab the solrconfig from the container:

```
docker create --name tmp1 solr:latest
docker cp tmp1:/opt/solr/server/solr/configsets/_default/conf/solrconfig.xml solrconfig.xml
docker rm tmp1
```

Modify it to load the dataimporthandler, and add a requesthandler configuration for it:

```
ed solrconfig.xml <<'EOM'
$
?requestHandler>
a
<requestHandler name="/dataimport" class="org.apache.solr.handler.dataimport.DataImportHandler">
<lst name="defaults">
  <str name="config">data-config.xml</str>
</lst>
</requestHandler>
.

1
/regex="solr-velocity-
a
  <lib dir="${solr.install.dir:../../../..}/dist/" regex="solr-dataimporthandler-\d.*\.jar" />
  <lib dir="${solr.install.dir:../../../..}/contrib/dataimporthandler/lib" regex=".*\.jar" />
.
w
q
EOM
```

so that we end up with:
```
--- solrconfig.xml.orig	2019-06-22 12:34:11.000000000 +0100
+++ solrconfig.xml	2019-06-22 13:57:05.000000000 +0100
@@ -83,6 +83,8 @@
 
   <lib dir="${solr.install.dir:../../../..}/contrib/velocity/lib" regex=".*\.jar" />
   <lib dir="${solr.install.dir:../../../..}/dist/" regex="solr-velocity-\d.*\.jar" />
+  <lib dir="${solr.install.dir:../../../..}/dist/" regex="solr-dataimporthandler-\d.*\.jar" />
+  <lib dir="${solr.install.dir:../../../..}/contrib/dataimporthandler/lib" regex=".*\.jar" />
   <lib dir="${solr.install.dir:../../../..}/dist/" regex="solr-ltr-\d.*\.jar" />
 
   <!-- an exact 'path' can be used instead of a 'dir' to specify a
@@ -1003,6 +1005,14 @@
       <str>elevator</str>
     </arr>
   </requestHandler>
+<requestHandler name="/dataimport" class="org.apache.solr.handler.dataimport.DataImportHandler">
+<lst name="defaults">
+  <str name="config">data-config.xml</str>
+</lst>
+</requestHandler>
 
   <!-- Highlighting Component

```

To run:

```
docker run -it -p 8879:8983 \
  -v $PWD/solrconfig.xml:/opt/solr/server/solr/configsets/_default/conf/solrconfig.xml \
  -v $PWD/data-config.xml:/opt/solr/server/solr/configsets/_default/conf/data-config.xml \
  -v $PWD/postgresql-42.2.6.jar:/opt/solr/contrib/dataimporthandler/lib/postgresql-42.2.6.jar \
  solr:8 \
  solr-precreate core1
```

That's the core created.

Now go to  http://localhost:8879/solr/#/core1/dataimport and hit Execute to do a full import.
Check the Solr log; there should be no errors.
Then go to http://localhost:8879/solr/#/core1/query and do a query -- you should see docs.

So that is Solr importing from Postgres.


Now modify the data-config.xml to have the following variables:

```
<dataConfig>
  <dataSource type="JdbcDataSource" 
              driver="org.postgresql.Driver"
              url="${dataimporter.datasource.url}"
              user="${dataimporter.datasource.user}"
              password="${dataimporter.datasource.password}"/>
  <document>
    <entity name="id" 
            query="select id,name from city">
     <field column="id" name="id"/>
     <field column="name" name="name_str"/>
    </entity>
  </document>
</dataConfig>
```

and pass those properties (I use `-e SOLR_OPTS` rather than the `-a` though that would presumably work too):

```
docker run -it -p 8879:8983 \
  -v $PWD/solrconfig.xml:/opt/solr/server/solr/configsets/_default/conf/solrconfig.xml \
  -v $PWD/data-config.xml:/opt/solr/server/solr/configsets/_default/conf/data-config.xml \
  -v $PWD/postgresql-42.2.6.jar:/opt/solr/contrib/dataimporthandler/lib/postgresql-42.2.6.jar \
  -e SOLR_OPTS="-Ddataimporter.datasource.url=jdbc:postgresql://172.17.0.2/world -Ddataimporter.datasource.user=postgres -Ddataimporter.datasource.password=postgres" \
  solr:8 \
  solr-precreate core1
```

With that, I can still import from the DB. So try that.

I've added this to https://github.com/docker-solr/docker-solr-examples/tree/master/dih-postgres

**Comment by KiaKK** [comment]
Hi @makuk66 , thank you for the detailed responses!! We are trying to run solr on different servers and each solr instance will pull the data from different sql server for DIH, so instead of hard coding the DB info in each of the data_config.xml, I am trying to pass in it while running the container.
 
I tried to add -a "-Ddatasource.user=REAUSERNAME" and actually it did work (oops..). I was checking the Configuration file displayed on Solr Admin and it didn't get replaced, but if I hit Execute it did trigger a full import without any issues.

But your suggestion of using -e SOLR_OPTS is even better for us. I just tried and its working!!

About solr 8 upgrade -- Our cores' configs are all in ConfigNoData/solr. With solr 7.7.1, two cores are correctly generated after running the command to start the container. But by using 8.1.1, with the same command and dockerfile in the original post, the cores are not created automatically.
